app/cdap/services/datasource/index.js

```diff
@@ -54,7 +55,7 @@ export default class Datasource {
       }
 
       genericResponseHandlers.forEach((handler) => handler(data));
-
+      redirectOnError(data.response);
```

Comment:
Chances of losing work is very minimal as we have polling (every 10secs) in all pages and in case of CMEK error it won't be short.

Other approach would be:
As you pointed out, through error handler we can trigger (via eventEmitter) a modal/dialog. But product is inclined towards full page error.